VSCODE-89 field names completion

alenakhineika opened the pull request:
<!-- Ticket number and a general summary of your changes in the Title above -->
<!-- e.g. COMPASS-1111: updates ace editor width in agg pipeline view -->

<!--- The following fields are not obligatory. Use your best judgement on what you think is applicable to the work you've done -->

## Description
<!--- Describe your changes in detail -->
<!--- If applicable, describe (or illustrate) architecture flow -->
- Parse text from the playground editor to build AST and provide fields completion.
- Use AST to identify `db.collection.` substrings and send only them to the shell parser, instead of all text from the editor. We have to do it because the current shell implementation can't properly handle situations with `.` symbols in string literals and comments.

<img width="1184" alt="Screenshot 2020-04-23 at 13 56 01" src="https://user-images.githubusercontent.com/16307679/80096652-33c81680-856a-11ea-8c24-375a9cb49a6d.png">

When a user types a query in a playground, we check:
- if the current symbol belongs to the object key `db.collection.find({ j})`
- if there is a text before the current symbol  that includes `use("db")` and `db.collection.` calls
- if we have cached fields for the `"db.collection"` namespace
- if cached fields have items that start from the letter user typed

If there is a match we show field suggestion. If not:

- we search for a document within the current namespace
- if found one parse the schema to get fields
- cache fields to avoid extra parsing and improve performance (we clear cache each time after running a playground.)

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it's updating a dependancy, link to the Pull Request that originally introduced the fix -->
- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

Anemy commented on src/test/suite/language/mongoDBService.test.ts:
```diff
@@ -1,64 +1,364 @@
+      expect(findCompletion).to.be.undefined;
+    });
+
+    test('do not provide fields completion if has wrong collection', async () => {
```
Should the provideCompletionItems be returning anything here? Maybe we can add not returning anything that as another `expect`.

Anemy commented on src/language/mongoDBService.ts:
```diff
@@ -181,4 +200,233 @@ export default class MongoDBService {
+        }
+      });
+    } catch (error) {
+      this._connection.console.log(`ESPRIMA error: ${util.inspect(error)}`);
```
Nice - I think it's important we're tolerant of errors and keep running - we're not quite sure what the users might have in their editor and it could be pretty error prone.

alenakhineika replied:
Yeah we try to be very alive in any case.

alenakhineika commented on src/language/server.ts:
```diff
@@ -225,6 +225,18 @@ connection.onDidChangeWatchedFiles((_change) => {
 });
 
+// Execute the entire playground script.
+connection.onRequest('executeAll', (codeToEvaluate, token) => {
```
Do you mean to have something like `serverCommands.ECECUTE_ALL` instead of `'executeAll'`?

Anemy replied:
Yup - I'm thinking it can help save against possible bugs.

alenakhineika commented on src/test/helper/playgroundHelper.ts:
```diff
@@ -0,0 +1,69 @@
+//   } catch (error) {
+//     return Promise.reject();
+//   }
+// }
```
We have couple of commented helpers for the language server tests. If you don't mind I would keep them for a while. They can be useful for next iterations, and I remove them later when I will be sure we don't need them.

alenakhineika commented on src/test/suite/language/mongoDBService.test.ts:
```diff
@@ -1,64 +1,364 @@
 const chai = require('chai');
 const expect = chai.expect;
 
+const INCREASED_TEST_TIMEOUT = 5000;
```
All tests that call mongodh evaluator. They work okayish for Linux and Mac builds, but for windows, they take about 3500 ms.

<img width="970" alt="Screenshot 2020-04-24 at 21 24 12" src="https://user-images.githubusercontent.com/16307679/80249379-04043600-8672-11ea-942f-f5b347c9c29d.png">

<img width="972" alt="Screenshot 2020-04-24 at 21 23 30" src="https://user-images.githubusercontent.com/16307679/80249385-06ff2680-8672-11ea-98f6-e212e9e9458e.png">

alenakhineika commented:
@Anemy thank you very much for finding all these issues!! before Max has found them 😆 Fixed them and also added more tests to cover these use cases. Here is a ticket for DB and collection names completion: https://jira.mongodb.org/browse/VSCODE-93

Anemy commented:
@alenakhineika Nice looks good - small q - we currently have the `CompletionItemKind` for the autocompletions set to 'Text' https://code.visualstudio.com/api/references/vscode-api#CompletionItemKind Maybe we should use `Field`? Maybe more of a design choice. For reference it's the wrench looking icon next to `price` I think:
<img width="295" alt="Screen Shot 2020-04-25 at 1 50 12 AM" src="https://user-images.githubusercontent.com/1791149/80265353-979c2d80-8697-11ea-8957-f65d6c9c4cd6.png">

alenakhineika commented:
@Anemy Hah I used `{ limit: 1, projection: { _id: 0 } }` options for find documents query. Removed projection to return `_id` as well. Also added server commands enum and changed kind to `Field` for fields completion and `Method` for shell symbols completion.
